File: test/bench/java/rbtree.java

enum Color {
  Red,
  Black
}


interface FoldFun {
  int Apply(int k, boolean v, int acc);
}

class Tree {
  Color color;
  Tree left;
  int  key;
  boolean val;
  Tree right;

  Tree( Color c, Tree l, int k, boolean v, Tree r)  {
    color = c;
    left = l;
    key = k;
    val = v;
    right = r;
  }

  static Tree Node( Color c, Tree l, int k, boolean v, Tree r)  {
    return new Tree(c,l,k,v,r);
  }

  static boolean isRed( Tree t ) {
    return (t != null && t.color == Color.Red);
  }

  static Tree balanceRight( int kv, boolean vv, Tree t, Tree n ) {
    if (n == null) {
      return null;
    }
    else if (n.left != null && n.left.color == Color.Red) {
      //case let .Node(_, .Node(.Red, l, kx, vx, r1), ky, vy, r2):
      //  return .Node(.Red, .Node(.Black, l, kx, vx, r1), ky, vy, .Node(.Black, r2, kv, vv, t))
      Tree l = n.left;
      return Node( Color.Red, Node( Color.Black, l.left, l.key, l.val, l.right), n.key, n.val, Node(Color.Black, n.right, kv, vv, t));
    }
    else if (n.right != null && n.right.color == Color.Red) {
      //case let .Node(_, l1, ky, vy, .Node(.Red, l2, kx, vx, r)):
      //  return .Node(.Red, .Node(.Black, l1, ky, vy, l2), kx, vx, .Node(.Black, r, kv, vv, t))
      Tree r = n.right;
      return Node( Color.Red, Node( Color.Black, n.left, n.key, n.val, r.left), r.key, r.val, Node(Color.Black, r.right, kv, vv, t));
    }
    else {
      //case let .Node(_, l, ky, vy, r):
      //  return .Node(.Black, .Node(.Red, l, ky, vy, r), kv, vv, t)
      return Node(Color.Black, Node(Color.Red, n.left, n.key, n.val, n.right), kv, vv, t);
    }
  }

  static Tree balanceLeft( Tree t, int kv, boolean vv, Tree n ) {
    if (n == null) {
      return null;
    }
    else if (n.left != null && n.left.color == Color.Red) {
      //case let .Node(_, .Node(.Red, l, kx1, vx1, r1), ky, vy, r2):
      //  return .Node(.Red, .Node(.Black, t, kv, vv, l), kx1, vx1, .Node(.Black, r1, ky, vy, r2))
      Tree l = n.left;
      return Node( Color.Red, Node( Color.Black, t, kv, vv, l.left), l.key, l.val, Node(Color.Black, l.right, n.key, n.val, n.right));
    }
    else if (n.right != null && n.right.color == Color.Red) {
      //case let .Node(_, l1, ky, vy, .Node(.Red, l2, kx2, vx2, r2)):
      //  return .Node(.Red, .Node(.Black, t, kv, vv, l1), ky, vy, .Node(.Black, l2, kx2, vx2, r2))
      Tree r = n.right;
      return Node( Color.Red, Node( Color.Black, t, kv, vv, n.left), n.key, n.val, Node(Color.Black, r.left, r.key, r.val, r.right));
    }
    else {
      //case let .Node (_, l, ky, vy, r):
      //  return .Node(.Black, t, kv, vv, .Node(.Red, l, ky, vy, r))
      return Node(Color.Black, t, kv, vv, Node(Color.Red, n.left, n.key, n.val, n.right));
    }
  }

  static Tree ins(Tree t, int kx, boolean vx ) {
    if (t==null) {
      return Node(Color.Red, null, kx, vx, null);
    }
    else if (t.color == Color.Red) {
      //case let .Node(.Red, a, ky, vy, b):
      if (kx < t.key) {
        return Node(Color.Red, ins(t.left, kx, vx), t.key, t.val, t.right);
      } else if (t.key == kx) {
        return Node(Color.Red, t.left, kx, vx, t.right);
      } else {
        return Node(Color.Red, t.left, t.key, t.val, ins(t.right, kx, vx));
      }
    }
    else { // t.color == Black
      if (kx < t.key) {
        if (isRed(t.left)) {
          return balanceRight(t.key, t.val, t.right, ins(t.left, kx, vx));
        } else {
          return Node(Color.Black, ins(t.left, kx, vx), t.key, t.val, t.right);
        }
      } else if (kx == t.key) {
        return Node(Color.Black, t.left, kx, vx, t.right);
      } else {
        if (isRed(t.right)) {
          return balanceLeft(t.left, t.key, t.val, ins(t.right, kx, vx));
        } else {
          return Node(Color.Black, t.left, t.key, t.val, ins(t.right, kx, vx));
        }
      }
    }
  }

  static Tree setBlack( Tree t ) {
    if (t == null) return t;
    return Node(Color.Black, t.left, t.key, t.val, t.right);
  }

  static Tree insert (Tree t, int k, boolean v) {
    if (isRed(t)) {
      return setBlack(ins(t, k, v));
    } else {
      return ins(t, k, v);
    }
  }

  static int Fold( FoldFun f, Tree t, int acc ) {
    while(t != null) {
      acc = Fold(f,t.left,acc);
      acc = f.Apply(t.key,t.val,acc);
      t = t.right;
    }
    return acc;
  }

}


public class rbtree
{
  static Tree mkMap( int n ) {
    Tree t = null;
    while(n > 0) {
      n--;
      t = Tree.insert(t, n, (n%10)==0);
    }
    return t;
  }

  static int Test(int n ) {
    Tree t = mkMap(n);
    return Tree.Fold( (k,v,acc) -> { return (v ? acc + 1 : acc); }, t, 0);
  }

  public static void main(String args[])
  {
      System.out.println( Test(4200000) );
  }
}

File: test/bench/haskell/rbtree.hs
-- Adapted from https://github.com/leanprover/lean4/blob/IFL19/tests/bench/rbmap.hs
-- Modified to be strict in the Tree fields
import System.Environment

data Color = Red | Black

data Tree k v
  = Leaf
  | Node !Color !(Tree k v) !k !v !(Tree k v)

fold :: (a -> b  -> c  -> c) -> Tree a b -> c  -> c
fold _ Leaf b               = b
fold f (Node _ l k v r)   b = fold f r (f k v (fold f l b))

balance1 :: Tree a b -> Tree a b -> Tree a b
balance1 (Node _ _ kv vv t) (Node _ (Node Red l kx vx r₁) ky vy r₂) = Node Red (Node Black l kx vx r₁) ky vy (Node Black r₂ kv vv t)
balance1 (Node _ _ kv vv t) (Node _ l₁ ky vy (Node Red l₂ kx vx r)) = Node Red (Node Black l₁ ky vy l₂) kx vx (Node Black r kv vv t)
balance1 (Node _ _ kv vv t) (Node _ l  ky vy r)                     = Node Black (Node Red l ky vy r) kv vv t
balance1 _                                                        _ = Leaf

balance2 :: Tree a b -> Tree a b -> Tree a b
balance2 (Node _ t kv vv _) (Node _ (Node Red l kx₁ vx₁ r₁) ky vy r₂)  = Node Red (Node Black t kv vv l) kx₁ vx₁ (Node Black r₁ ky vy r₂)
balance2 (Node _ t kv vv _) (Node _ l₁ ky vy (Node Red l₂ kx₂ vx₂ r₂)) = Node Red (Node Black t kv vv l₁) ky vy (Node Black l₂ kx₂ vx₂ r₂)
balance2 (Node _ t kv vv _) (Node _ l ky vy r)                         = Node Black t kv vv (Node Red l ky vy r)
balance2 _                                                        _    = Leaf

is_red :: Tree a b -> Bool
is_red (Node Red _ _ _ _) = True
is_red _                  = False

ins :: Ord a => Tree a b -> a -> b -> Tree a b
ins Leaf                 kx vx = Node Red Leaf kx vx Leaf
ins (Node Red l ky vy r) kx vx =
   (if kx < ky then Node Red (ins l kx vx) ky vy r
    else if ky < kx then Node Red l ky vy (ins r kx vx)
    else Node Red l kx vx r) -- Node Red l ky vy (ins r kx vx))
ins (Node Black l ky vy r) kx vx =
    if kx < ky then
      (if is_red l then balance1 (Node Black Leaf ky vy r) (ins l kx vx)
       else Node Black (ins l kx vx) ky vy r)
    else if ky < kx then
      (if is_red r then balance2 (Node Black l ky vy Leaf) (ins r kx vx)
       else Node Black l ky vy (ins r kx vx))
    else Node Black l kx vx r

set_black :: Tree a b -> Tree a b
set_black (Node _ l k v r) = Node Black l k v r
set_black e                = e

insert t k v =
  if is_red t then set_black (ins t k v)
  else ins t k v

type Map = Tree Int Bool

mk_Map_aux :: Int -> Map -> Map
mk_Map_aux 0 m = m
mk_Map_aux n m = let n' = n-1 in mk_Map_aux n' (insert m n' (n' `mod` 10 == 0))

mk_Map n = mk_Map_aux n Leaf

main = do
  -- [arg] <- getArgs
  -- let n :: Int = read arg
  let n = 4200000
  let m = mk_Map n
  let v = fold (\_ v r -> if v then r + 1 else r) m 0
  print v
